contracts/interchain-token-service/src/abi.rs

```diff
@@ -0,0 +1,503 @@
+                        .then(|| Bytes::from_slice(env, decoded.minter.as_ref())),
+                })
+            }
+            _ => return Err(MessageError::InvalidMessageType),
+        };
+
+        Ok(message)
+    }
+}
+
+impl HubMessage {
+    pub fn abi_encode(self, env: &Env) -> Bytes {
+        let msg = match self {
+            Self::SendToHub(types::SendToHub {
+                destination_chain,
+                message,
+            }) => SendToHub {
+                messageType: U256::from(types::MessageType::SendToHub as u32),
+                destination_chain: soroban_string_to_string(destination_chain),
+                message: message.abi_encode(env).to_alloc_vec().into(),
+            }
+            .abi_encode_params(),
+            Self::ReceiveFromHub(types::ReceiveFromHub {
+                source_chain,
+                message,
+            }) => ReceiveFromHub {
+                messageType: U256::from(types::MessageType::ReceiveFromHub as u32),
+                source_chain: soroban_string_to_string(source_chain),
+                message: message.abi_encode(env).to_alloc_vec().into(),
+            }
+            .abi_encode_params(),
+        };
+        Bytes::from_slice(env, &msg)
+    }
+
+    pub fn abi_decode(env: &Env, payload: &Bytes) -> Result<Self, MessageError> {
+        ensure!(payload.len() >= 32, MessageError::InsufficientMessageLength);
+
+        let binding = payload.to_alloc_vec();
+        let payload_array = binding.as_slice();
+
+        let message_type = MessageType::abi_decode(&payload_array[0..32], true)
+            .map_err(|_| MessageError::InvalidMessageType)?;
+
+        let message = match message_type {
+            MessageType::SendToHub => {
+                let decoded = SendToHub::abi_decode_params(payload_array, true)
+                    .map_err(|_| MessageError::AbiDecodeFailed)?;
+
+                Self::SendToHub(types::SendToHub {
+                    destination_chain: String::from_str(env, &decoded.destination_chain),
+                    message: Message::abi_decode(
+                        env,
+                        &Bytes::from_slice(env, decoded.message.as_ref()),
+                    )?,
+                })
+            }
+            MessageType::ReceiveFromHub => {
+                let decoded = ReceiveFromHub::abi_decode_params(payload_array, true)
+                    .map_err(|_| MessageError::AbiDecodeFailed)?;
+
+                Self::ReceiveFromHub(types::ReceiveFromHub {
+                    source_chain: String::from_str(env, &decoded.source_chain),
+                    message: Message::abi_decode(
+                        env,
+                        &Bytes::from_slice(env, decoded.message.as_ref()),
+                    )?,
+                })
+            }
+            _ => return Err(MessageError::InvalidMessageType),
+        };
+
+        Ok(message)
+    }
+}
+
+fn soroban_string_to_string(soroban_string: String) -> StdString {
```

Comment:
does soroban_sdk not provide a From conversion between their own string types?

Reply:
@milapsheth  There is a `to_string()` method provided, but this is where we run into compiler issues when building the wasm, since implementing it requires using the `Display` trait, which means pulling in `std::fmt`. Is implementing `Display` for Soroban's `String` type preferable to a custom method?

Reply:
We can't implement a foreign crate on a foreign type. I think soroban sdk should have gated this under testutils feature since it has an std dependency. Maybe report the issue to them?